### ISSUE TITLE
Farmhash-modern throws webassembly error in Nextjs

### ISSUE THREAD

**Issue by B-uchi** [firebase/firebase-admin-node#2627]
### [REQUIRED] Step 2: Describe your environment

  * Operating System version: Linux Ubuntu 24.04
  * Firebase Admin SDK version: v12.2.0
  * Firebase Product: Admin auth
  * Node.js version: v20.11.1
  * NPM version: v10.2.4

### [REQUIRED] Step 3: Describe the problem
I'm trying to configure firebase-admin SDK on my nextjs + TS project. Whenever i try to invoke any SDK function, i get a webAssembly error. In my case, i am trying to configure a middleware for the server side api, and when i call the verifyIdToken method it throws the error. I'll paste it below, i also noticed that the affected dependency `farmhash-modern` was just replaced in this version of firebase.

#### Steps to reproduce:
The error occured when i called the `verifyIdToken()` method in my `middleware.ts` file.

#### Relevant Code:
This is the middleware file.
```
import { auth } from "./lib/config/firebaseConfig";
import { NextResponse } from "next/server";
import { NextRequest } from "next/server";

interface ExtendedNextRequest extends NextRequest {
  uid?: string;
}

export async function middleware(req: ExtendedNextRequest, res: NextResponse) {
  try {
    let token = req.headers.get("Authorization");
    if (!token)
      return NextResponse.json({ message: "Access Denied" }, { status: 403 });
    if (token.startsWith("Bearer ")) {
      token = token.split(" ")[1];
    }

    let decodedToken = await auth.verifyIdToken(token);
    let uid = decodedToken.uid;
    req.uid = uid;

    return NextResponse.next();
  } catch (error: any) {
    console.log(error.errorInfo);
    return NextResponse.json({ message: "Invalid Token" }, { status: 401 });
  }
}
```
#### This is the error output
```
⨯ ./node_modules/farmhash-modern/bin/bundler/farmhash_modern_bg.wasm
Module parse failed: Unexpected character '' (1:0)
The module seem to be a WebAssembly module, but module is not flagged as WebAssembly module for webpack.
BREAKING CHANGE: Since webpack 5 WebAssembly is not enabled by default and flagged as experimental feature.
You need to enable one of the WebAssembly experiments via 'experiments.asyncWebAssembly: true' (based on async modules) or 'experiments.syncWebAssembly: true' (like webpack 4, deprecated).
For files that transpile to WebAssembly, make sure to set the module type in the 'module.rules' section of the config (e. g. 'type: "webassembly/async"').
(Source code omitted for this binary file)

Import trace for requested module:
./node_modules/farmhash-modern/bin/bundler/farmhash_modern_bg.wasm
./node_modules/farmhash-modern/bin/bundler/farmhash_modern.js
./node_modules/farmhash-modern/lib/browser.js
./node_modules/firebase-admin/lib/remote-config/condition-evaluator-internal.js
./node_modules/firebase-admin/lib/remote-config/remote-config.js
./node_modules/firebase-admin/lib/app/firebase-namespace.js
./node_modules/firebase-admin/lib/default-namespace.js
./node_modules/firebase-admin/lib/index.js
./lib/config/firebaseConfig.ts
```

This question is also on [stackoverflow](https://stackoverflow.com/questions/78729243/firebase-admin-dependency-farmhash-modern-throws-webassembly-error)

**Comment by lahirumaramba** [comment]
Does configuring webpack by setting `asyncWebAssembly` to `true` in your `next.config.js` work for you?
Something similar to the following:

```js
module.exports = {
  webpack: (config, { buildId, dev, isServer, defaultLoaders, webpack }) => {
    config.experiments = {
      asyncWebAssembly: true,
    };
    return config;
  },
};
```

**Comment by Keegan-lee** [comment]
I'm getting the same issue.. This code does seem to "fix" the issue, but `src/middleware` is now failing (hanging) to compile.

> Does configuring webpack by setting `asyncWebAssembly` to `true` in your `next.config.js` work for you? Something similar to the following:
> 
> ```js
> module.exports = {
>   webpack: (config, { buildId, dev, isServer, defaultLoaders, webpack }) => {
>     config.experiments = {
>       asyncWebAssembly: true,
>     };
>     return config;
>   },
> };
> ```

**Comment by raymelon** [comment]
Experiencing the same issue.

**Comment by matija2209** [comment]
Downgrading to 12.0.0 worked for me as well.

**Comment by B-uchi** [comment]
> Does configuring webpack by setting `asyncWebAssembly` to `true` in your `next.config.js` work for you?
> Something similar to the following:
> 
> ```js
> module.exports = {
>   webpack: (config, { buildId, dev, isServer, defaultLoaders, webpack }) => {
>     config.experiments = {
>       asyncWebAssembly: true,
>     };
>     return config;
>   },
> };
> ```

I also tried this, but as @Keegan-lee said `src/middleware` doesn't compile. It displays the compiling middleware message and stays there forever.

**Comment by nicolopadovan** [comment]
I am experiencing the same exact issue.
Adding the `asyncWebAssembly` flag indeed causes an indefinite compiling for `middleware`.
Switching back to `^12.0.0` doesn't work either.

**Workaround**
For now, I have moved all the pages that needed that middleware into a specific route (can also be an anonymous route).
I am using the `layout` file as a kind of middleware: before displaying any data it runs a server action, where the issue is not yielded.
This workaround produces an error when hot reloading, but seems to work fine otherwise.

Also, using `^12.0.0` issued an error regarding the usage of the `os` module, which may be a clue on where to debug this issue.

**Comment by Lucasch1** [comment]
Am having the same problem here, searched the best i could and it seems that there is no direct solution yet...

**Comment by nicolopadovan** [comment]
> Am having the same problem here, searched the best i could and it seems that there is no direct solution yet...
> 
> The best that i had saw is utilize another lib called next-firebase-auth-edge (https://github.com/awinogrodzki/next-firebase-auth-edge)

I am guessing that Firebase Auth cannot be used on Edge runtime due to the need for some Node-only modules.
FWIW I suggest to use this library with **extreme caution** due to how it manipulates and generates tokens and other security-related Firebase Auth features.

**Comment by hassanzadeh** [comment]
Hey Guys,
I'm having the same issue, but I completely replaced auth part of firebase-admin with next-firebase-auth-edge, only using admin part and firestore, should I still be getting this error?

**Comment by nicolopadovan** [comment]
> Hey Guys, I'm having the same issue, but I completely replaced auth part of firebase-admin with next-firebase-auth-edge, only using admin part and firestore, should I still be getting this error?

You shouldn’t get the error, the package should be able to smoothly run on all NextJS environments.

[I however renew my suggestion of being cautious when using the package](https://github.com/firebase/firebase-admin-node/issues/2627#issuecomment-2302065812)

**Comment by hassanzadeh** [comment]
> > Hey Guys, I'm having the same issue, but I completely replaced auth part of firebase-admin with next-firebase-auth-edge, only using admin part and firestore, should I still be getting this error?
> 
> You shouldn’t get the error, the package should be able to smoothly run on all NextJS environments.
> 
> [I however renew my suggestion of being cautious when using the package](https://github.com/firebase/firebase-admin-node/issues/2627#issuecomment-2302065812)

Are you sure? I see plenty of people complaining about the same thing (that it won't work when using middleware). I'm using firestore and admin (but not auth) from firebase-admin and it leads to that error.

**Comment by nicolopadovan** [comment]
> > > Hey Guys, I'm having the same issue, but I completely replaced auth part of firebase-admin with next-firebase-auth-edge, only using admin part and firestore, should I still be getting this error?
> > 
> > 
> > You shouldn’t get the error, the package should be able to smoothly run on all NextJS environments.
> > [I however renew my suggestion of being cautious when using the package](https://github.com/firebase/firebase-admin-node/issues/2627#issuecomment-2302065812)
> 
> Are you sure? I see plenty of people complaining about the same thing (that it won't work when using middleware). I'm using firestore and admin (but not auth) from firebase-admin and it leads to that error.

How are you importing the library?

Also, yes I am sure, as per the documentation of the package:

```
The official firebase-admin library depends heavily on Node.js’s internal crypto library, which isn’t available in [Next.js Edge Runtime](https://nextjs.org/docs/api-reference/edge-runtime).

This library solves that problem by handling the creation and verification of [Custom ID Tokens](https://firebase.google.com/docs/auth/admin/verify-id-tokens) using the Web Crypto API, which works in Edge runtimes
```

**Comment by hassanzadeh** [comment]
updated my response above.

**Comment by brouza** [comment]
As pointed out earlier, Firebase seems to work only in the Node server.
My workaround was to create an API endpoint with call API/session/route and export a response as JSON from the @auth session, then call the auth API in the middleware. 
This works well for me, but I'm unsure about its security implications.

// /api/auth/session
import { auth } from "@/auth"


export async function GET(_request: Request){
    const session = await auth()
    if(!session) return Response.json(null, { status: 401 })
    return new Response(JSON.stringify(session), {
        status: 200,
    })
}

// middleware.ts
export async function middleware(_request: NextRequest) {
  const res = await fetch(new URL('/api/auth/session', _request.url), {
    method: 'GET',
    credentials: 'include',
    headers: {
      'Content-Type': 'application/json',
    },
  })
  const schema = z.custom<Session | null>()
  const isAuthenticated = schema.parse(await res.json())
  
  console.log(isAuthenticated)

  const respond = NextResponse.next()

  return respond
}

**Comment by nicolopadovan** [comment]
First of all, I suggest that you all try to format the code in the thread correctly. [You can do so via backticks](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code).
Also, try to keep the thread clean! Don't turn it into a StackOverflow question! :D 

@hassanzadeh I can't test your code right now, but the issue might lie in the fact that you are still using `firebase-admin` within `firebase-init.ts`, which you are then importing in your middleware.

@brouza Seems like an okay workaround as far as I'm concerned; you could even try using a server action to abstract away the REST paradigm. I would at least try to secure the endpoint, however, even though it's a server-to-server communication. There might be more security implications as well.

I am as well awaiting an official response regarding this issue.

**Comment by hassanzadeh** [comment]
Thanks @nicolopadovan ,
I updated my response to fix the backticks issue.
I'm not sure I understood your point, I'm importing firebase-init, but all I'm using there is admin and db, but not auth, is that a problem?

**Comment by hassanzadeh** [comment]
> As pointed out earlier, Firebase seems to work only in the Node server. My workaround was to create an API endpoint with call API/session/route and export a response as JSON from the @auth session, then call the auth API in the middleware. This works well for me, but I'm unsure about its security implications.
> 
> // /api/auth/session import { auth } from "@/auth"
> 
> export async function GET(_request: Request){ const session = await auth() if(!session) return Response.json(null, { status: 401 }) return new Response(JSON.stringify(session), { status: 200, }) }
> 
> // middleware.ts export async function middleware(_request: NextRequest) { const res = await fetch(new URL('/api/auth/session', _request.url), { method: 'GET', credentials: 'include', headers: { 'Content-Type': 'application/json', }, }) const schema = z.custom<Session | null>() const isAuthenticated = schema.parse(await res.json())
> 
> console.log(isAuthenticated)
> 
> const respond = NextResponse.next()
> 
> return respond }

Thanks, I first need to know why my current code is not working especially that I'm not using firebase-admin auht at all.

**Comment by hassanzadeh** [comment]
Ok, I think you are right.
Thanks

**Comment by rwlp** [comment]
Use webassembly in middlewares is not common practice. Using native code means you are applying a complex solution to improve efficiency. Middleware in Next.js is designed for handling requests and responses, but your approach may be misusing it.

In the case of Firebase, it's best to separate the methods into distinct files, such as firebaseServerSide.ts and firebaseClientSide.ts. Again, middleware in Next.js does not support running WebAssembly, and it’s generally a bad practice to run this type of code there (since firebase-admin uses WebAssembly).

My recommendation is to enhance your login and authentication processes. Consider implementing straightforward checks, such as validating tokens with JWT or other libraries, or even using libraries like next-auth for this purpose.

Another important observation is that if you are using firebase-admin on the client side(or import methods that use firebase-admin), it is considered a bad practice and poses significant security risks, which will likely lead to similar problems.

**Comment by B-uchi** [comment]
> Use webassembly in middlewares is not common practice. Using native code means you are applying a complex solution to improve efficiency. Middleware in Next.js is designed for handling requests and responses, but your approach may be misusing it.
> 
> In the case of Firebase, it's best to separate the methods into distinct files, such as firebaseServerSide.ts and firebaseClientSide.ts. Again, middleware in Next.js does not support running WebAssembly, and it’s generally a bad practice to run this type of code there (since firebase-admin uses WebAssembly).
> 
> My recommendation is to enhance your login and authentication processes. Consider implementing straightforward checks, such as validating tokens with JWT or other libraries, or even using libraries like next-auth for this purpose.
> 
> Another important observation is that if you are using firebase-admin on the client side(or import methods that use firebase-admin), it is considered a bad practice and poses significant security risks, which will likely lead to similar problems.

Thanks for the insight. I did some research and understand the issue better.

**Comment by Pavel-Husakouski** [comment]
Same error when using 12.6.0

**Comment by lahirumaramba** [comment]
We removed the dependency on `farmhash-modern` in the latest release `v13`. This should be now fixed in the latest release. Let us know if you are still running into issues. Thanks!

https://github.com/firebase/firebase-admin-node/pull/2760

**Comment by nicholas-masini** [comment]
> We removed the dependency on `farmhash-modern` in the latest release `v13`. This should be now fixed in the latest release. Let us know if you are still running into issues. Thanks!
> 
> [#2760](https://github.com/firebase/firebase-admin-node/pull/2760)

Still running into the same issue from my end. Exact same scenario as OP, I have a middleware which is checking if the JWT is valid, `firebase-admin` is `v13` however still getting the same error

**Comment by nicolopadovan** [comment]
With [NextJS 15.5](https://nextjs.org/blog/next-15-5), the problem should be fixed, as NodeJS Middleware is now stable

```
export const config = {
  runtime: 'nodejs', // Now stable!
};
```

Haven't been able to test this myself - and as @lahirumaramba said, the issue should've been already fixed with #2760 - but it's worth mentioning for future reference / issues.

**Comment by B-uchi** [comment]
> With [NextJS 15.5](https://nextjs.org/blog/next-15-5), the problem should be fixed, as NodeJS Middleware is now stable
> 
> ```
> export const config = {
>   runtime: 'nodejs', // Now stable!
> };
> ```
> 
> Haven't been able to test this myself - and as @lahirumaramba said, the issue should've been already fixed with #2760 - but it's worth mentioning for future reference / issues.

Thanks!
